scipy/optimize/_simulated_annealing.py

```diff
@@ -0,0 +1,559 @@
+"""
+SDA: A Simulated Dual Annealing global optimization algorithm
+"""
+from __future__ import division, print_function, absolute_import
+
+import numpy as np
+from scipy.optimize import OptimizeResult
+from scipy.optimize import minimize
+from scipy.special import gammaln
+from scipy._lib._util import check_random_state
+
+__all__ = ['simulated_annealing']
+BIG_VALUE = 1e16
+
+
+class VisitingDistribution(object):
+    """
+    Class used to generate new coordinates based on the distorted
+    Cauchy-Lorentz distribution. Depending on the steps within the Markov
+    chain, the class implements the strategy for generating new location
+    changes.
+    """
+    tail_limit = 1.e8
+    min_visit_bound = 1.e-10
+
+    def __init__(self, lb, ub, visiting_param, rand_state):
+        self.visiting_param = visiting_param
+        self.rand_state = rand_state
+        self.lower = lb
+        self.upper = ub
+        self.bound_range = ub - lb
+        self.x_gauss = None
+        self.s_gauss = 0
+        self.root_gauss = None
+
+    def visiting(self, x, step, temperature):
+        dim = x.size
+        if step < dim:
+            # Changing all coordinates with a new visting value
+            visits = np.array([self.visit_fn(
+                temperature) for _ in range(dim)])
+            upper_sample = self.rand_state.random_sample()
+            lower_sample = self.rand_state.random_sample()
+            visits[visits > self.tail_limit] = self.tail_limit * upper_sample
+            visits[visits < -self.tail_limit] = -self.tail_limit * lower_sample
+            x_visit = visits + x
+            a = x_visit - self.lower
+            b = np.fmod(a, self.bound_range) + self.bound_range
+            x_visit = np.fmod(b, self.bound_range) + self.lower
+            x_visit[np.fabs(
+                x_visit - self.lower) < self.min_visit_bound] += 1.e-10
+        else:
+            # Changing only one coordinate at a time based on Markov chain step
+            x_visit = np.copy(x)
+            visit = self.visit_fn(temperature)
+            if visit > self.tail_limit:
+                visit = self.tail_limit * self.rand_state.random_sample()
+            elif visit < -self.tail_limit:
+                visit = -self.tail_limit * self.rand_state.random_sample()
+            index = step - dim
+            x_visit[index] = visit + x[index]
+            a = x_visit[index] - self.lower[index]
+            b = np.fmod(a, self.bound_range[index]) + self.bound_range[index]
+            x_visit[index] = np.fmod(b, self.bound_range[
+                index]) + self.lower[index]
+            if np.fabs(x_visit[index] - self.lower[
+                    index]) < self.min_visit_bound:
+                x_visit[index] += self.min_visit_bound
+        return x_visit
+
+    def visit_fn(self, temperature):
+        factor1 = np.exp(np.log(temperature) / (self.visiting_param - 1.0))
+        factor2 = np.exp((4.0 - self.visiting_param) * np.log(
```

Comment:
1. I think I have already said that, but I think you need to define the values such as `4.0`, `2.0`, `0.5`... as constants with more explicative names.
2. I don't like the use of the name factor1, 2, 3,... for variables, maybe you should come up with more meaningful names.

Reply:
Hi @antonior92, thanks for the comments. This function, variable names and constants are provided by Constantino Tsallis in the paper (Tsallis C, Stariolo DA (1996). "Generalized Simulated Annealing." Physica A, 233, 395-406.). There is no particular meaning for it and it is the formula expression (look at page 405 of the paper).

Reply:
Ok, so maybe include a comment citing the formula from the original paper. Something like:

```
# Formula X from p. 405 of reference [2]
```

Reply:
Commented as such. Thanks!